### PR TITLE
feat(layers): Migrate RandomChoice and RandomApply layers from keras-cv to keras

### PR DESCRIPTION
**Pull request by harshaljanjani**
- Added `RandomChoice` and `RandomApply` layers to `keras.layers.preprocessing.image_preprocessing`
- Implemented necessary transform methods (`transform_images`, `transform_labels`, `transform_bounding_boxes`, `transform_segmentation_masks`) to comply with `BaseImagePreprocessingLayer`
- Updated tests to ensure compatibility with keras and fix failing test cases
- Added support for batchwise processing, auto-vectorization, and random seed control
- Addresses #20727

### CODE REVIEW

**Comment by harshaljanjani** [comment]
I ran the `random_apply_test.py` and `random_choice_test.py` unit tests locally across all backends, and they seemed to work. In any case, the backends other than TensorFlow seem to be unhappy about the change. Would love to know if there's an obvious breaking change I've overlooked, will get back to the PR as soon as time permits, thanks!

File: keras/src/layers/preprocessing/image_preprocessing/random_apply.py [hunk 1]
```diff
@@ -0,0 +1,192 @@
...
+        self.built = True
+
+    def _get_should_augment(self, inputs):
+        input_shape = tf.shape(inputs)
+        
+        if self.batchwise:
+            return self._rate > tf.random.uniform(shape=(), seed=self.seed)
+        
+        batch_size = input_shape[0]
+        random_values = tf.random.uniform(shape=(batch_size,), seed=self.seed)
```
Comment (innat) [inline keras/src/layers/preprocessing/image_preprocessing/random_apply.py:115]: You may need to update tf ops with backend agnostic ops. So that, thse layers can be added to model architecture (gpu leverage for data augmentation).

Comment (harshaljanjani) [reply keras/src/layers/preprocessing/image_preprocessing/random_apply.py]: Thanks for calling that out! Swapped ops to backend-agnostic alternatives wherever I could :)

File: keras/src/layers/preprocessing/image_preprocessing/random_apply.py [hunk 2]
```diff
@@ -0,0 +1,262 @@
...
+import keras.src.random as random
+from keras.src import ops
+from keras.src.api_export import keras_export
+from keras.src.layers.preprocessing.image_preprocessing.base_image_preprocessing_layer import (  # noqa: E501
+    BaseImagePreprocessingLayer,
+)
+
+
+@keras_export("keras.layers.RandomApply")
+class RandomApply(BaseImagePreprocessingLayer):
```
Comment (innat) [inline keras/src/layers/preprocessing/image_preprocessing/random_apply.py:25]: These layers (random-apply/choice) are not specificly image processing layer, rather than general. So, it might be better to subclass from `keras.layers.Layer`, no?

File: keras/src/layers/preprocessing/image_preprocessing/random_apply_test.py [hunk 3]
```diff
@@ -0,0 +1,141 @@
...
+        self.assertEqual(num_zero_inputs, 0)
+        self.assertLess(num_zero_outputs, batch_size)
+        self.assertGreater(num_zero_outputs, 0)
+
+    # TODO: find the root cause of the error
+    @pytest.mark.skipif(
+        K.backend() == "jax",
+        reason="""
+        jax.errors.UnexpectedTracerError:
+        Encountered an unexpected tracer
```
Comment (innat) [inline keras/src/layers/preprocessing/image_preprocessing/random_apply_test.py:82]: This may give some pointer. 
https://github.com/keras-team/keras/issues/20724

File: keras/src/layers/preprocessing/image_preprocessing/random_apply.py (hunk 2 above)
Comment (harshaljanjani) [reply keras/src/layers/preprocessing/image_preprocessing/random_apply.py:25]: Right, I had a hunch about this, which is why I mentioned it in the original issue thread. I'll wait for a review from @fchollet and make the change if needed. Thanks for pointing it out!

File: keras/src/layers/preprocessing/image_preprocessing/random_apply_test.py (hunk 3 above)
Comment (harshaljanjani) [reply keras/src/layers/preprocessing/image_preprocessing/random_apply_test.py:82]: Thanks for the lead, tried fixing the tracer leak by precomputing seeds and avoiding lambdas as suggested in `#20724`. Got 5/6 tests passing as before, and the `RandomFlip` with `batchwise=True` one still fails. The issue seems to be a tad bit deeper and might relate to how `RandomFlip` interacts with `RandomChoice` and as a result the PRNG key somehow escapes the scope of the transform. Tried `JAX_CHECK_TRACER_LEAKS=1`, but that didn’t provide much clarity either, will get back to it when time permits!

File: keras/src/layers/preprocessing/image_preprocessing/random_apply.py [hunk 4]
```diff
@@ -0,0 +1,279 @@
+# Copyright 2022 The KerasCV Authors
```
Comment (fchollet) [inline keras/src/layers/preprocessing/image_preprocessing/random_apply.py:1]: Please remove header

File: keras/src/layers/preprocessing/image_preprocessing/random_apply.py [hunk 5]
```diff
@@ -0,0 +1,279 @@
...
+
+    **Example:**
+    ```python
+    # Create a layer that zeroes out all pixels in an image
+    zero_out = keras.layers.Lambda(lambda x: 0 * x)
+
+    # Create a batch of random 2x2 images
+    images = tf.random.uniform(shape=(5, 2, 2, 1))
+    print(images[..., 0])
+    # <tf.Tensor: shape=(5, 2, 2), dtype=float32, numpy=
```
Comment (fchollet) [inline keras/src/layers/preprocessing/image_preprocessing/random_apply.py:43]: Code examples should not include TF mentions. Also it needs to be much shorter

File: keras/src/layers/preprocessing/image_preprocessing/random_apply.py [hunk 6]
```diff
@@ -0,0 +1,279 @@
...
+            individually (False). When True, the layer is either applied to
+            entire batch or not at all. When False, the layer is applied
+            independently to each input in the batch. Defaults to False.
+        auto_vectorize: A boolean, for whether to use vectorized operations for
+            batched inputs. Can improve performance but may not work with XLA.
+            Defaults to False.
+        seed: An integer, used to seed the random number generator for
+            reproducibility. Defaults to None.
+
+    **Call Arguments:**
```
Comment (fchollet) [inline keras/src/layers/preprocessing/image_preprocessing/random_apply.py:102]: Use e.g. `Call arguments:`, `Returns:`

File: keras/src/layers/preprocessing/image_preprocessing/random_apply.py [hunk 7]
```diff
@@ -0,0 +1,279 @@
...
+            }
+        else:
+            return self._call_single(inputs)
+
+    def _call_single(self, inputs):
+        inputs_rank = len(inputs.shape)
+        is_single_sample = ops.equal(inputs_rank, 3)
+        is_batch = ops.equal(inputs_rank, 4)
+
+        if K.backend() == "jax":
```
Comment (fchollet) [inline keras/src/layers/preprocessing/image_preprocessing/random_apply.py:221]: This is not the way to do it -- please check out other preprocessing layers for an example of how to handle the seed generator and seed argument

**Comment by harshaljanjani** [comment]
@fchollet, would love your thoughts on these changes whenever you get a chance, thanks!

File: keras/src/layers/preprocessing/image_preprocessing/random_apply.py [hunk 8]
```diff
@@ -0,0 +1,161 @@
...
+    Output: A tensor with the same shape and dtype as the input, with the
+        transformation layer applied to selected inputs.
+    """
+
+    def __init__(
+        self,
+        layer,
+        rate=0.5,
+        batchwise=False,
+        auto_vectorize=False,
```
Comment (fchollet) [inline keras/src/layers/preprocessing/image_preprocessing/random_apply.py:40]: What's this? it's not documented.

File: keras/src/layers/preprocessing/image_preprocessing/random_apply.py (hunk 2 above)
Comment (fchollet) [reply keras/src/layers/preprocessing/image_preprocessing/random_apply.py:10]: Given the purpose of the layer, its currently location and parent class makes sense.

File: keras/src/layers/preprocessing/image_preprocessing/random_apply.py [hunk 9]
```diff
@@ -0,0 +1,161 @@
...
+from keras.src.layers.preprocessing.image_preprocessing.base_image_preprocessing_layer import (  # noqa: E501
+    BaseImagePreprocessingLayer,
+)
+from keras.src.random.seed_generator import SeedGenerator
+
+
+@keras_export("keras.layers.RandomApply")
+class RandomApply(BaseImagePreprocessingLayer):
+    """A preprocessing layer that randomly applies a specified layer during
+    training.
```
Comment (fchollet) [inline keras/src/layers/preprocessing/image_preprocessing/random_apply.py:12]: Shorten to fit on one line.

File: keras/src/layers/preprocessing/image_preprocessing/random_apply.py [hunk 10]
```diff
@@ -0,0 +1,161 @@
...
+    This layer randomly applies a given transformation layer to inputs based on
+    the `rate` parameter. It is useful for stochastic data augmentation to
+    improve model robustness. At inference time, the output is identical to
+    the input. Call the layer with `training=True` to enable random application.
+
+    Args:
+        layer: A `keras.Layer` to apply. The layer must not modify input shape.
+        rate: Float between 0.0 and 1.0, representing the probability of
+            applying the layer. Defaults to 0.5.
+        batchwise: Boolean. If True, the decision to apply the layer is made for
```
Comment (fchollet) [inline keras/src/layers/preprocessing/image_preprocessing/random_apply.py:23]: Please use backticks around keywords like True and False

File: keras/src/layers/preprocessing/image_preprocessing/random_apply.py [hunk 11]
```diff
@@ -0,0 +1,161 @@
...
+
+    def transform_labels(self, labels, transformation, training=True):
+        if not training or transformation is None:
+            return labels
+        return self.transform_images(labels, transformation, training)
+
+    def transform_bounding_boxes(self, bboxes, transformation, training=True):
+        if not training or transformation is None:
+            return bboxes
+        return self.transform_images(bboxes, transformation, training)
```
Comment (fchollet) [inline keras/src/layers/preprocessing/image_preprocessing/random_apply.py:141]: This would be incorrect

**Review by fchollet — COMMENTED** [review]
Thanks for the PR!

@innat do you want to review?

**Comment by innat** [comment]
Sure, I'll review it. Thanks for tagging me.

File: keras/src/layers/preprocessing/image_preprocessing/random_choice.py [hunk 12]
```diff
@@ -0,0 +1,175 @@
...
+        super().__init__(**kwargs)
+        self.layers = layers
+        self.auto_vectorize = auto_vectorize
+        self.batchwise = batchwise
+        self.seed = seed
+        self.generator = SeedGenerator(seed)
+        self._convert_input_args = False
+        self._allow_non_tensor_positional_args = True
+
+    def get_random_transformation(self, data, training=True, seed=None):
```
Comment (innat) [inline keras/src/layers/preprocessing/image_preprocessing/random_choice.py:59]: @harshaljanjani 
In the [original implementation](https://github.com/keras-team/keras-cv/blob/94b0a551d03ee9de3d81663e2c1a680cb113f7f1/keras_cv/src/layers/preprocessing/random_choice.py#L24), there are no transformation functions present. Given the purpose of `RandomChoice`, it shouldn’t include any transformation methods since its sole responsibility is to randomly select one layer from multiple options and use it. Are there any challenges in adhering strictly to the original implementation?

File: keras/src/layers/preprocessing/image_preprocessing/random_apply.py (hunk 8 above)
Comment (harshaljanjani) [reply keras/src/layers/preprocessing/image_preprocessing/random_apply.py:40]: Ah, apologies for the oversight! Actually, `auto_vectorize` is an unimplemented parameter that slipped in during development in `RandomChoice` and `RandomApply` since I was trying to stick to the original implementation to the tee. However, with the final changes, I've removed it from the docstring, the `__init__` signature and `config` as well, as it turns out that even in the `keras-cv` implementation, it served no real purpose. Thanks for catching that!

File: keras/src/layers/preprocessing/image_preprocessing/random_apply.py (hunk 11 above)
Comment (harshaljanjani) [reply keras/src/layers/preprocessing/image_preprocessing/random_apply.py:141]: Thanks for pointing this out! You're absolutely right. Masks, labels, and bounding boxes shouldn't be subjected to the same transforms as the images (or at least that's what I gathered from this comment :) ). I've fixed this by following the patterns used in the majority of preprocessing layers, like `RandomContrast` and `RandomHue`, where these methods simply return the input unchanged.

File: keras/src/layers/preprocessing/image_preprocessing/random_choice.py (hunk 12 above)
Comment (harshaljanjani) [reply keras/src/layers/preprocessing/image_preprocessing/random_choice.py:59]: Absolutely not! Actually I was told to follow the patterns used in writing the code for the other layers (specifically for seed generation and passing the seeds as an argument I'm aware), but I noticed in most of the random layers this method was present so I added it, but touché I do see your point. I'll make the change.

**Comment by harshaljanjani** [comment]
Thanks for the reviews, @fchollet and @innat. I've tried to incorporate the suggested changes, will check the results of the test cases as soon as I have the opportunity!

**Comment by innat** [comment]
@fchollet 
My review is pending, it might require approval.

**Comment by harshaljanjani** [comment]
@fchollet please do look into the changes, thanks!

**Comment by fchollet** [comment]
> My review is pending, it might require approval.

I am not seeing anything to approve? I was under the impression anyone could add a review.

> @fchollet please do look into the changes, thanks!

Looking now!

**Comment by harshaljanjani** [comment]
Hi @fchollet,
Just wanted to kindly follow up to see if you had a chance to review the changes, please let me know if there's anything else I need to address, thanks!

**Comment by fchollet** [comment]
> Tests / Check the code format (pull_request) Failing after 1m

Please fix the code format by running `sh shell/format.sh`. If any errors remain, fix them manually.

File: keras/src/layers/preprocessing/image_preprocessing/random_apply.py [hunk 13]
```diff
@@ -0,0 +1,151 @@
...
+        flattened = ops.reshape(images, (num_batches, -1))
+        any_nonzero = ops.any(ops.not_equal(flattened, 0), axis=1)
+        num_non_zero_batches = ops.sum(ops.cast(any_nonzero, dtype="int32"))
+        return num_batches - num_non_zero_batches
+
+    def transform_labels(self, labels, transformation, training=True):
+        return labels
+
+    def transform_bounding_boxes(self, bboxes, transformation, training=True):
+        return bboxes
```
Comment (fchollet) [inline keras/src/layers/preprocessing/image_preprocessing/random_apply.py:134]: Seems incorrect -- the bboxes should be transformed by the inner layer if applied

File: keras/src/layers/preprocessing/image_preprocessing/random_apply.py [hunk 14]
```diff
@@ -0,0 +1,151 @@
...
+
+    def transform_segmentation_masks(
```
Comment (fchollet) [inline keras/src/layers/preprocessing/image_preprocessing/random_apply.py:136]: Same here

File: keras/src/layers/preprocessing/image_preprocessing/random_apply.py [hunk 15]
```diff
@@ -0,0 +1,151 @@
...
+            )
+        else:
+            augmented = self._layer(images)
+
+        return self.backend.numpy.where(should_augment, augmented, images)
+
+    def call(self, inputs):
+        if isinstance(inputs, dict):
+            return {
+                key: self._call_single(input_tensor)
```
Comment (fchollet) [inline keras/src/layers/preprocessing/image_preprocessing/random_apply.py:111]: Seems incorrect -- it should be something like:

```
if apply:
   return self._layer(inputs)
```

This is independent of whether inputs are a dict or not

**Comment by fchollet** [comment]
I am not even sure we should be brining over these layers in the first place, to be honest.

**Comment by harshaljanjani** [comment]
> I am not even sure we should be brining over these layers in the first place, to be honest.

In that case, please let me know if you'd like me to proceed with implementing these fixes or if the PR should be closed without merging. If these layers aren't needed in the codebase, it might make more sense to close the PR rather than investing time in fixing them.

**Comment by harshaljanjani** [comment]
Hi, is there a reason this PR is not moving forward? Please do clarify, thanks!

**Comment by fchollet** [comment]
Sorry for the delay -- as a conclusion, I decided we would not be bringing these 2 specific layers into Keras core. Thank you for your understanding.
